godot-core/src/builtin/array.rs

```diff
@@ -185,61 +282,189 @@ impl Array {
+    }
+
+    /// Sets the type of the inner array. Can only be called once, directly after creation.
+    fn set_typed(&mut self) {
+        debug_assert!(self.is_empty());
+        debug_assert!(!self.type_info().is_typed());
+        let type_info = TypeInfo::new::<T>();
+        if type_info.is_typed() {
+            let script = Variant::nil();
+            unsafe {
+                (interface_fn!(array_set_typed))(
+                    self.sys(),
+                    type_info.variant_type.sys(),
+                    type_info.class_name.string_sys(),
+                    script.var_sys(),
+                );
+            }
+        }
+    }
+}
+
+impl<T: VariantMetadata + FromVariant> TypedArray<T> {
+    /// Returns an iterator over the elements of the `TypedArray`. Note that this takes the array
+    /// by reference but returns its elements by value, since they are internally converted from
+    /// `Variant`.
+    ///
+    /// Notice that it's possible to modify the `TypedArray` through another reference while
+    /// iterating over it. This will not result in unsoundness or crashes, but will cause the
+    /// iterator to behave in an unspecified way.
+    pub fn iter_shared(&self) -> TypedArrayIterator<'_, T> {
+        TypedArrayIterator {
+            array: self,
+            next_idx: 0,
+        }
     }
 
-    /// Returns the value at the specified index as a `Variant`. To convert to a specific type, use
-    /// the available conversion methods on `Variant`, such as [`Variant::try_to`] or
-    /// [`Variant::to`].
+    /// Returns the value at the specified index.
     ///
     /// # Panics
     ///
     /// If `index` is out of bounds.
-    pub fn get(&self, index: usize) -> Variant {
+    pub fn get(&self, index: usize) -> T {
         let ptr = self.ptr(index);
-        // SAFETY: `ptr` just verified that the index is not out of bounds.
-        unsafe { (*ptr).clone() }
+        // SAFETY: `ptr()` just verified that the index is not out of bounds.
+        let variant = unsafe { &*ptr };
+        T::try_from_variant(variant).unwrap()
     }
 
     /// Returns the first element in the array, or `None` if the array is empty. Equivalent of
     /// `front()` in GDScript.
-    pub fn first(&self) -> Option<Variant> {
-        (!self.is_empty()).then(|| self.as_inner().front())
+    pub fn first(&self) -> Option<T> {
+        (!self.is_empty()).then(|| {
+            let variant = self.as_inner().front();
+            T::try_from_variant(&variant).unwrap()
+        })
     }
 
     /// Returns the last element in the array, or `None` if the array is empty. Equivalent of
     /// `back()` in GDScript.
-    pub fn last(&self) -> Option<Variant> {
-        (!self.is_empty()).then(|| self.as_inner().back())
+    pub fn last(&self) -> Option<T> {
+        (!self.is_empty()).then(|| {
+            let variant = self.as_inner().back();
+            T::try_from_variant(&variant).unwrap()
```

Comment:
`T::from_variant`

This occurs quite a few times in this file, maybe search for "try_from_variant".
Now that the `convenience` feature is removed on master, `from_variant` should be accessible everywhere.

Reply:
Done. I didn't mind the extra `unwrap()` since it clearly signals "if this ever triggers, it's a bug".

Reply:
> Done. I didn't mind the extra unwrap() since it clearly signals "if this ever triggers, it's a bug".

That's pretty much exactly the rationale behind these panicking methods as well, I just wrote it down today 🙂 

https://github.com/godot-rust/gdextension/blob/eb5d43b3e021e1eb1b67661d91318660cd54794c/godot/src/lib.rs#L72-L75

I agree it may be less obvious to general Rust developers that aren't familiar with the library, but I hope these design documents help explain the trade-offs this library makes 💡